api/server/prom_zip_collector.go

```diff
@@ -0,0 +1,80 @@
+package server
+
+import (
+	"github.com/fnproject/fn/api/agent"
+	"github.com/openzipkin/zipkin-go-opentracing"
+	"github.com/openzipkin/zipkin-go-opentracing/thrift/gen-go/zipkincore"
+	"github.com/prometheus/client_golang/prometheus"
+	"strings"
+)
+
+// Each span name is published as a separate Histogram metric
+// Using metric names of the form fn_span_<span-name>_duration_seconds
+var histogramVecMap = make(map[string]*prometheus.HistogramVec)
+
+// Return the HistogramVec corresponding to the specified spanName.
+// If a HistogramVec does not already exist for specified spanName then one is created and configured with the specified labels
+// otherwise the labels parameter is ignored.
+func getHistogramVecForSpanName(spanName string, labels []string) *prometheus.HistogramVec {
+	thisHistogramVec, found := histogramVecMap[spanName]
+	if !found {
+		thisHistogramVec = prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name: "fn_span_" + spanName + "_duration_seconds",
+				Help: "Span " + spanName + " duration, by span name",
+			},
+			labels,
+		)
+		histogramVecMap[spanName] = thisHistogramVec
+		prometheus.MustRegister(thisHistogramVec)
+	}
+	return thisHistogramVec
+}
+
+// PrometheusCollector is a custom Collector
+// which sends ZipKin traces to Prometheus
+type PrometheusCollector struct {
+	a agent.Agent
+}
+
+// NewPrometheusCollector returns a new PrometheusCollector
+func NewPrometheusCollector(agent agent.Agent) (zipkintracer.Collector, error) {
+	pc := &PrometheusCollector{}
+	pc.a = agent
+	return pc, nil
+}
+
+// PrometheusCollector implements Collector.
+func (pc PrometheusCollector) Collect(span *zipkincore.Span) error {
+
+	// extract any label values from the span
+	labelKeys, labelValueMap := getLabels(span)
+
+	getHistogramVecForSpanName(span.GetName(), labelKeys).With(labelValueMap).Observe(float64(span.GetDuration()) / 1000000)
```

Comment:
would be great to get millisecond resolution, as many (most?) stats are likely in that range or smaller. would prefer to see a constant of `time.Second` instead of magic number `1000000` too (easier to understand without needing a comment)

Reply:
Changed  `float64(span.GetDuration()) / 1000000` to `span.GetDuration().Seconds()`, which is of course much better. 

I chose seconds because the Prometheus documentation [best practices for naming metrics and labels](https://prometheus.io/docs/practices/naming) that metrics should use base units (i.e. seconds in this case). Is that a good enough reason?

Reply:
Changed  `float64(span.GetDuration()) / 1000000` to `(time.Duration(span.GetDuration())*time.Microsecond).Seconds()`. 

I chose seconds because the Prometheus documentation [best practices for naming metrics and labels](https://prometheus.io/docs/practices/naming) that metrics should use "base units" (seconds for time units).

Reply:
`time.Duration.Seconds()` returns a float64 we should be able to work with just fine 👍 

since we're multiplying by `time.Microsecond` I guess `span.GetDuration()` returns a value in microseconds that we want to convert to nanoseconds to be a time.Duration and then convert to seconds? is this logic correct?